datafusion/src/optimizer/filter_push_down.rs

```diff
@@ -253,33 +176,107 @@ fn split_members<'a>(predicate: &'a Expr, predicates: &mut Vec<&'a Expr>) {
     }
 }
 
+// For a given JOIN logical plan, determine whether each side of the join is preserved.
+// We say a join side is preserved if the join returns all or a subset of the rows from
+// the relevant side - i.e. the side of the join cannot provide nulls. Returns a tuple
+// of booleans - (left_preserved, right_preserved).
+fn lr_is_preserved(plan: &LogicalPlan) -> (bool, bool) {
+    match plan {
+        LogicalPlan::Join(Join { join_type, .. }) => match join_type {
+            JoinType::Inner => (true, true),
+            JoinType::Left => (true, false),
+            JoinType::Right => (false, true),
+            JoinType::Full => (false, false),
```

Comment:
Both sides are preserved in a ` FULL` join (as in this should probably be (`true`, `true`)

Reply:
Ah - my impression was that `FULL` was a `FULL OUTER` join - i.e. both sides are nullable/non-preserved? Is this not right? I might have mixed it up with `CrossJoin`.

Reply:
`Full` is `FULL OUTER` join... I think in that case the inputs are both preserved (as in you can't push filters down through the join if it isn't for that column)

Though that doesn't seem consistent with `JoinType::Inner`

Maybe 
```rust
            JoinType::Inner => (false, false),
``` 
as predicates *can* be pushed down both sides there 🤔

Reply:
Maybe whether a side is `nullable` is the better distinction here? I see your point about both sides of the `FULL` join being preserved, but you still can't push filters because they can provide extra null rows.

Reply:
I think there are two different things making this confusing:

1.  for predicates that are `null preserving` (like `t1 <= 5`) in the `WHERE` clause it is actually fine to create a predicate like `t2 < 5` even for outer joins (try it with postgres, you'll find that predicate filters out any rows that didn't have matches on either side so it is ok)
2. For predicates that are not `null preserving` (like `t1 is NULL`) you *CAN'T* push predicates down (which I think is the core of this PR)

To make things even more confusing, the rules of when you can push / not push are different if the predicate is in the `ON` clause)

Also I don't think DataFusion currently tracks the `null preserving` property for `Exprs` -- so the safe (correct thing) is to treat all predicates like they may not be `null preserving` and not create/push them for outer joins. A follow on exercise would be to allow more predicates to be pushed down by checking `null preserving`